### DISCUSSION TITLE
How to add modules/packages completion and auto-import

### DISCUSSION THREAD

**Discussion by zaynOm** [LazyVim/LazyVim#3685] [Q&A] [answered]
I'm having an issue with modules/packages completion, like if I want to use sys in python I would have to manually import it, if i type sys it does not show up in the auto completion popup.

**Comment by dpetka2001** [comment]
This has to do with the LSP server that you're using. See this [comment](https://github.com/hrsh7th/nvim-cmp/issues/426#issuecomment-1185144017) in an issue in `nvim-cmp` repo. Quoting
> The current behavior was copied from the TypeScript language server which limits its auto-imports to only those symbols and modules that have been imported somewhere else in the program.

It also has a link to an issue at pyright repo.

**Reply by zaynOm** [reply]
I'm trying to change the default LSP maybe I can get better experience, I did the following but I doesn't seam to work.

lua/plugins/lsp.lua
```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      pylsp = {
        enabled = lsp == "pylsp",
      },
    },
  },
}
```

lua/config/options.lua
```lua
vim.g.lazyvim_python_lsp = "pylsp"
```

**Reply by dpetka2001** [reply, marked as answer]
Your lspconfig should be like this
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        pylsp = {
          settings = {
            pylsp = {
              plugins = {
                rope_autoimport = {
                  enabled = true,
                },
              },
            },
          },
        },
      },
    },
  },
```
But this will pull in all `pylsp` dependencies that are listed [here](https://pypi.org/project/python-lsp-server/). Personally, I just install `pylsp` in every virtual environment and only additionally install `rope` which is used for auto-completion as I prefer to rely on ruff instead of the tools that are builtin with `pylsp` metapackage.

**Reply by zaynOm** [reply]
I have a virtual environment setup for nvim, can I use it to install them

**Reply by dpetka2001** [reply]
Not sure. I haven't tried that. I just create a virtual environement for every project. Try it to see if it works out for you and share your results :smile:

**Reply by zaynOm** [reply]
thx, Its working perfectly, the only thing though is when i check `LSPInfo` I see pyright and basepyright running should I just uninstall them using mason

**Reply by zaynOm** [reply]
> Not sure. I haven't tried that. I just create a virtual environement for every project. Try it to see if it works out for you and share your results 😄

It did work, I just installed them in a virtual environment linked with nvim, so i don't have to install them each time

**Reply by dpetka2001** [reply]
Yes you should do it yourself, since there's currently no automatic procedure for uninstalling mason packages as far as i know (except if there's some other plugin that does this that i'm not aware of).

**Reply by dpetka2001** [reply]
> It did work, I just installed them in a virtual environment linked with nvim, so i don't have to install them each time

Cool!! That's good to know. I'll also try that. Thnx for letting me know :smile:

**Reply by dpetka2001** [reply]
Can you change your configuration to this
```lua
 {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        pylsp = {
		  mason = false,
          settings = {
            pylsp = {
              plugins = {
                rope_autoimport = {
                  enabled = true,
                },
              },
            },
          },
        },
      },
    },
  },
```
Because without `mason = false` it will still download and use `pylsp` from Mason and not your virtual environment (and also remove `python-lsp-server` from Mason). This way you'll know for sure if it works from your virtual environment linked to your nvim.

**Reply by zaynOm** [reply]
after doing what you suggested it this warning appeared when I opened a python file:
`Client pylsp quit with exit code 127 and signal 0. Check log for errors: /home/hox/.local/state/nvim/lsp.log`

and the completion isn't working any more, although `LSPInfo` shows that `pylsp` is active:
```lua
 Language client log: /home/hox/.local/state/nvim/lsp.log
 Detected filetype:   python
 
 1 client(s) attached to this buffer: 
 
 Client: ruff_lsp (id: 2, bufnr: [11])
 	filetypes:       python
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /home/hox/.local/share/nvim/mason/bin/ruff-lsp
 
 Other clients that match the filetype: python
 
 Config: pylsp
 	filetypes:         python
 	root directory:    /home/hox/Projects/test
 	cmd:               /home/hox/.pyenv/shims/pylsp
 	cmd is executable: true
 	autostart:         true
 	custom handlers:   
 ```

**Reply by dpetka2001** [reply]
Yep, so it doesn't work from the Neovim linked environment. It has to be installed in the virtual environment of the project. So, either install it in each virtual environment that you want or just let Mason install the whole meta-package and use that. I do the former because I don't want all the other dependencies of `pylsp` since I just use `rope` for auto-completion, but that's just my personal preference. Do whatever fits your needs best.

**Reply by zaynOm** [reply]
I appreciate your help

**Reply by zaynOm** [reply]
I checked the logs and it seams that the pylsp command is not found:
```
[START][2024-06-15 19:10:45] LSP logging initiated
[ERROR][2024-06-15 19:10:45] .../vim/lsp/rpc.lua:770	"rpc"	"/home/hox/.local/share/nvim/mason/bin/lua-language-server"	"stderr"	"NMD, WSM `endLine > #rows` ?\nstack traceback:\n\tscript/text-merger.lua:72: in upvalue 'mergeRows'\n\tscript/text-merger.lua:106: in function 'text-merger'\n\tscript/provider/provider.lua:305: in function <script/provider/provider.lua:294>\n\t[C]: in function 'xpcall'\n\tscript/proto/proto.lua:200: in function <script/proto/proto.lua:175>\n"
[START][2024-06-17 15:42:05] LSP logging initiated
[ERROR][2024-06-17 15:42:05] .../vim/lsp/rpc.lua:770	"rpc"	"/home/hox/.pyenv/shims/pylsp"	"stderr"	"pyenv: pylsp: command not found\n"
[ERROR][2024-06-17 15:42:05] .../vim/lsp/rpc.lua:770	"rpc"	"/home/hox/.pyenv/shims/pylsp"	"stderr"	"\nThe `pylsp' command exists in these Python versions:\n"
[ERROR][2024-06-17 15:42:05] .../vim/lsp/rpc.lua:770	"rpc"	"/home/hox/.pyenv/shims/pylsp"	"stderr"	"  nvim-venv\n"
[ERROR][2024-06-17 15:42:05] .../vim/lsp/rpc.lua:770	"rpc"	"/home/hox/.pyenv/shims/pylsp"	"stderr"	"\nNote: See 'pyenv help global' for tips on allowing both\n      python2 and python3 to be found.\n"
```
but after I activated the nvim virtual env it worked, but if it can not access these command installed in this venv why does nvim ask me to install pynvim in it.

**Reply by dpetka2001** [reply]
No idea to be honest. I've never done this. I always open Neovim from inside the activated virtual environment and Neovim picks up the correct Python version in the corresponding environment. Maybe I'm missing something else, but I honestly don't know since I've never tried it.

**Reply by dpetka2001** [reply]
So, in `nvim-lspconfig` where you're defining the `pylsp` server, you can also pass a `cmd` key that holds the path to the command (see [here](https://github.com/neovim/nvim-lspconfig/blob/bd7c76375a511994c9ca8d69441f134dc10ae3bd/lua/lspconfig/server_configurations/pylsp.lua#L5)). Can you try that to see if you don't need to activate the nvim environment any more?

**Reply by zaynOm** [reply]
that's what I was trying, but it doesn't seam to work, im sure im missing something, it gives me this:
```/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:800: Spawning language server with cmd: `{ "~/.pyenv/versions/nvim-venv/bin/pylsp" }` failed. The language server is either not installed, missing from PATH, or not executable.```

I wanted to install pylsp system wide but i can't using pip, so I tried using the pacman but rope isn't available there. if you're not using arch try to install them system wide using your package manager and you would not need to install them each time in a venv.

**Comment by CarterFendley** [comment]
Adding some information about [ty](https://github.com/astral-sh/ty) which is made by the same company who develops `ruff`. They mention [here](https://github.com/astral-sh/ruff/issues/19490#issuecomment-3103665997) that `ruff` is not built to be a full fledged language server and that `ty` does this more complex cross-module actions. As noted [here](https://docs.astral.sh/ty/editors/) it works with neovim!

It was also work with a simple change to `lua/config/options.lua` and was automatically installed by LazyVim.
```lau
vim.g.lazyvim_python_lsp = "ty"
```

Then I disabled ruff because it was annoying me too in `lua/plugins/`.
> Note: I originally had `ty = {}` here but works with it commented out just fine.
```lua
{
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      -- ty = {},
      ruff = { enabled = false },
    },
  },
},
```
